plugin/plugins/filesystem/nodes/root.go

```diff
@@ -0,0 +1,139 @@
+package fsnodes
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/djdv/p9/p9"
+	cid "github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log"
+	coreiface "github.com/ipfs/interface-go-ipfs-core"
+	corepath "github.com/ipfs/interface-go-ipfs-core/path"
+	"github.com/multiformats/go-multihash"
+)
+
+var _ p9.File = (*RootIndex)(nil)
+
+func newRootPath(path string) corepath.Resolved {
+	return rootNode(path)
+}
+
+type rootNode string
+
+func (rn rootNode) String() string { return string(rn) }
+func (rootNode) Namespace() string { return nRoot }
+func (rootNode) Mutable() bool     { return true }
+func (rootNode) IsValid() error    { return nil } //TODO: should return ENotImpl
+func (rn rootNode) Cid() cid.Cid {
+	prefix := cid.V1Builder{Codec: cid.DagCBOR, MhType: multihash.BLAKE2B_MIN}
+	c, err := prefix.Sum([]byte(rn))
+	if err != nil {
+		panic(err) //invalid root
+	}
+	return c
+}
+func (rootNode) Root() cid.Cid { //TODO: this should probably reference a package variable set during init `rootCid`
+	prefix := cid.V1Builder{Codec: cid.DagCBOR, MhType: multihash.BLAKE2B_MIN}
+	c, err := prefix.Sum([]byte("/"))
+	if err != nil {
+		panic(err) //invalid root
+	}
+	return c
+}
+func (rootNode) Remainder() string { return "" }
+
+// RootIndex is a virtual directory file system, that maps a set of filesystem implementations to a hierarchy
+// Currently: "/":RootIndex, "/ipfs":PinFS, "/ipfs/*:IPFS
+type RootIndex struct {
+	IPFSBase
+	subsystems []p9.Dirent
+}
+
+// RootAttacher constructs the default RootIndex file system, providing a means to Attach() to it
+func RootAttacher(ctx context.Context, core coreiface.CoreAPI) *RootIndex {
+	ri := &RootIndex{
+		IPFSBase: newIPFSBase(ctx, newRootPath("/"), p9.TypeDir,
```

Comment:
This probably shouldn't use `IPFSBase` as that appears to assume that we have a "resolved" path.

(unless I'm missing something).

Reply:
I think it's only like this for overlapping purposes. In `Walk` we attach to other filesystems that also construct an IPFSBase, so we derive a new one from the struct fields initialized here.

I'm going to see if there's a better way to handle all of that. Really we should only need the core and the context to initialize any other fs.

Reply:
We handle node initialization and deriving slightly better now.